tests/benchmarks/test_tpch_pyspark.py

```diff
@@ -0,0 +1,338 @@
+                            con.load_extension("httpfs")
+                            con.sql(
+                                f"""
+                                SET s3_region='us-east-2';
+                                SET s3_access_key_id='{creds.access_key}';
+                                SET s3_secret_access_key='{creds.secret_key}';
+                                """
+                            )
+                        if hasattr(module, "ddl"):
+                            con.sql(module.ddl)
+                        _ = con.execute(module.query).fetch_arrow_table()
+
+                    elif engine == "dask":
+                        dask(None)  # No client needed/used by dask queries
+
+                    elif engine == "pyspark":
+                        import pyspark as pyspark_
+
+                        from .tpch_pyspark_queries.utils import get_or_create_spark
+
+                        queries.utils.MASTER = "local[*]"
+
+                        # TODO: context env vars
+                        os.environ["SPARK_HOME"] = str(
+                            pathlib.Path(pyspark_.__file__).absolute().parent
+                        )
+                        os.environ["PYSPARK_PYTHON"] = sys.executable
+                        os.environ["PYSPARK_SUBMIT_ARGS"] = make_pyspark_submit_args(
+                            queries.utils.MASTER
+                        )
+
+                        spark = get_or_create_spark(f"query{query_num}")
+
+                        module = getattr(queries, f"q{query_num}")
+                        module.setup(spark)  # read spark tables query will select from
+
+                        if hasattr(module, "ddl"):
+                            spark.sql(module.ddl)  # might create temp view
+
+                        # scale1000 stored as timestamp[ns] which spark parquet
+                        # can't use natively.
+                        if ENABLED_DATASET == "scale 1000":
+                            module.query = fix_timestamp_ns_columns(module.query)
+
+                        q_final = spark.sql(module.query)  # benchmark query
+                        try:
+                            # trigger materialization of df
+                            _ = q_final.collect()
+                        finally:
+                            spark.catalog.clearCache()
+                            spark.sparkContext.stop()
+                            spark.stop()
+
+                _.cluster.adapt(minimum=1, maximum=1)
+                _()
+
+            func.__name__ = f"test_query_{query_num}"
+            return func
+
+        for query_num in range(1, 23):
+            # Limit to queries implemented in dask
+            test_tpch_dask = getattr(test_tpch, f"test_query_{query_num}", None)
+            if test_tpch_dask is not None:
+                test = make_test(test_tpch_dask, test_tpch_pyspark, query_num)
+                setattr(cls, test.__name__, test)
```

Comment:
I'll admit that I find these tests somewhat confusing.  I feel like it's maybe too DRY, and so hard to reason about maybe.  I wonder if a different approach would make sense where we had some fixtures to do some of the preparation, and then had simple tests written like `test_tpch/duckdb.py::test_q1` `test_tpch/duckdb::test_q2` and so on.  This structure seems a little more flat / unmagical to me and maybe easier for others (or at least me) to understand.

Reply:
> It was suggested to just get something working and not concern oneself too much with the implementation.

Understood, and please don't take my comment here as a directive from on-high, but rather as a perspective of another engineer.   Personally I default to flat architectures, especially when other people have to interact with them.  I find that things that do automation like this tend to not get updated because other people have difficulty understanding them.